cni/cmd/install/main.go

```diff
@@ -0,0 +1,316 @@
+	cniConfName := env.GetString("CNI_CONF_NAME", find_cni_conf_file(mountedCniNetDir))
+	serviceAccountPath := "/var/run/secrets/kubernetes.io/serviceaccount"
+
+	copyBinaries(cniConfName)
+	err := prepareKubeconfig(mountedCniNetDir, kubecfgName, serviceAccountPath)
+	if err != nil {
+		return err
+	}
+	err = prepareKumaCniConfig(mountedCniNetDir, hostCniNetDir, kubecfgName, serviceAccountPath, cniConfName, chainedCniPlugin)
+	if err != nil {
+		return err
+	}
+
+	shouldSleep, _ := env.GetBool("SLEEP", true)
+
+	for shouldSleep == true {
+		time.Sleep(time.Duration(cfgCheckInterval) * time.Second)
+		check_install(mountedCniNetDir)
+	}
+
+	return nil
+}
+
+func prepareKumaCniConfig(mountedCniNetDir, hostCniNetDir, kubecfgName, serviceAccountPath, cniConfName string, chainedCniPlugin bool) error {
+	rawConfig := env.GetString("CNI_NETWORK_CONFIG", "")
+	kubeconfigFilePath := hostCniNetDir + "/" + kubecfgName
+
+	config := strings.Replace(rawConfig, "__KUBECONFIG_FILEPATH__", kubeconfigFilePath, 1)
+	logger.Default.Debug("config after replace", zap.String("config", config))
+
+	serviceAccountToken, err := ioutil.ReadFile(serviceAccountPath + "/token")
+	if err != nil {
+		return err
+	}
+	config = strings.Replace(config, "__SERVICEACCOUNT_TOKEN__", string(serviceAccountToken), 1)
+
+	if chainedCniPlugin {
+		err := setupChainedPlugin(mountedCniNetDir, cniConfName, config)
+		if err != nil {
+			logger.Default.Error("unable to setup kuma cni as chained plugin", zap.Error(err))
+			return err
+		}
+	}
+
+	return nil
+}
+
+func setupChainedPlugin(mountedCniNetDir, cniConfName, kumaCniConfig string) error {
+	resolvedName := cniConfName
+	extension := filepath.Ext(cniConfName)
+	if !fileExists(mountedCniNetDir+"/"+cniConfName) && extension == ".conf" && fileExists(mountedCniNetDir+"/"+cniConfName+"list") {
+		resolvedName = cniConfName + "list"
+	}
+
+	if fileExists(mountedCniNetDir + "/" + resolvedName) {
+		hostCniConfig, err := ioutil.ReadFile(mountedCniNetDir + "/" + resolvedName)
+		if err != nil {
+			return err
+		}
+
+		marshaled, err := transformJsonConfig(kumaCniConfig, hostCniConfig)
+		if err != nil {
+			return err
+		}
+		logger.Default.Debug("resulting config", zap.String("config", string(marshaled)))
+
+		err = atomic.WriteFile(mountedCniNetDir+"/"+resolvedName, bytes.NewReader(marshaled))
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+	return nil
+}
+
+func transformJsonConfig(kumaCniConfig string, hostCniConfig []byte) ([]byte, error) {
+	queryString := `if has("type") then
+   .plugins = [.]
+   | del(.plugins[0].cniVersion)
+   | to_entries
+   | map(select(.key=="plugins"))
+   | from_entries
+   | .plugins += [` + kumaCniConfig + `]
+   | .name = "k8s-pod-network"
+   | .cniVersion = "0.3.0"
+else
+  del(.plugins[]? | select(.type == "kuma-cni"))
+  | .plugins += [` + kumaCniConfig + `]
+end`
+	// this is probably going to be rewritten to not use `jq` at all
+	query, err := gojq.Parse(queryString)
```

Comment:
Yep that'd be nice

Reply:
You might have already as part of a lib a struct that enables you to obtain a Plain Go obkject

Reply:
There is a [way to load](https://github.com/containernetworking/cni/blob/main/libcni/conf.go#L64) it but the library does not have a function to marshal it back (and the case does not match). So it's either jq or using `map[string]interface{}`